compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -509,14 +512,48 @@ convertClassDef env tycon = do
     let fields = zipExact labels (map sanitize fieldTypes)
         tconName = mkTypeCon [getOccText tycon]
         tsynName = mkTypeSyn [getOccText tycon]
+        newStyle = envLfVersion env `supports` featureTypeSynonyms
+            -- "new-style typeclasses" are type synonyms
+            -- "old-style typeclasses" were record types
         typeDef
-            | envLfVersion env `supports` featureTypeSynonyms =
-              -- Structs must have > 0 fields, therefore we simply make a typeclass a synonym for Unit
-              -- if it has no fields
-              defTypeSyn tsynName tyVars (if null fields then TUnit else TStruct fields)
+            | newStyle =
+                -- LF structs must have > 0 fields, therefore we define the
+                -- typeclass as a synonym for Unit if it has no fields.
+                defTypeSyn tsynName tyVars (if null fields then TUnit else TStruct fields)
             | otherwise = defDataType tconName tyVars (DataRecord fields)
 
-    pure [typeDef]
+    let funDeps = snd (classTvsFds cls)
+    funDeps' <- mapM (mapFunDepM (convTypeVarName env')) funDeps
+
+    let funDepTyVars = [(v, KStar) | (v, _) <- tyVars]
+            -- We use the the type variables as types in the fundep encoding,
+            -- not as whatever kind they were previously defined.
+        funDepName = ExprValName ("$fd" <> getOccText tycon)
```

Comment:
Dictionaries start with `$f`. Can this become a problem? Should we try to find some definitely unoccupied space, like starting with `$$`?

Reply:
Oh, shoot. And the PR merged already... I'll open a new PR to rename this.

Fortunately I don't think a conflict is actually possible with `$fd` because dictionary functions names always start with `$fT` where `T` is the name of the class, and `T` can never start with a lowercase `d`.

Reply:
No, the dictionary function would be `$fFood` in that case.

Here's the PR: https://github.com/digital-asset/daml/pull/7701

Reply:
`$fFood`? That sounds yummy. 😋

Thanks for changing it.